018-function-execution-model-refactor/018-function-execution-model-refactor.md

```diff
@@ -0,0 +1,129 @@
+
+A separate RFC is currently being worked on for how runtime languages will be written. 
+
+#### API Endpoints
+This rewrite will add functionality to the API endpoints that are currently used by the functions worker. The executor will take over these endpoints while the rest will stay underneath the functions worker.
+
+`POST /v1/functions/{id}/execute` - Execute function API endpoint
+The standard execution endpoint for functions will be rewritten to use the executor.
+It will act like so:
+
+1. Check if the runtime is running, if not spin up the runtime.
+2. Make a request to the runtime passing the arguments recieve through the API request.
+3. If the request is asyncronous then return the API request now with a execution ID like it does now.
+4. When we receive a response back from the runtime we will do 1 of two things
+    If the request is synchronous:
+        - Wait for the response from the runtime and return it as the response to the API request.
+    If the request is asynchronous:
+        - Store the result for later use through the Get Execution API endpoint. 
+
+`PATCH /v1/functions/{id}/tag` - Update function tag API endpoint
+The tag update endpoint will also recieve a small rewrite for the request moving it into the executor.
+
+It will act like so:
+1. Extract the code tarball into the enviroment.
+2. Check if the server is running, if it is then stop the server.
+3. also check if the runtime exists, if it doesn't then create the runtime.
+4. Map the new code into the runtime environment using a volume
+5. Finally bring the runtime back up and start the server again.
+
+#### Watchdog
+Watchdog will be apart of the executor running in the background dealing with all the running runtimes and monitoring their health.
+Every X minutes (defaults will need to be determined) watchdog will call a health endpoint which will be built into all enviroment web servers.
+If the web server returns a '200' then the runtime is health and watchdog will continue onto the next runtime however if the runtime does not respond in a set time or responds incorrectly then the runtime will be restarted and watchdog will continue monitoring it.
+If after a certain amount of restarts the runtime still does not respond correctly then watchdog will log a error and inform the user that something is wrong
+with the code running in the runtime. It will also stop attempting to restart it to save on resources.
+
+Watchdog will also deal with cleaning up runtimes that haven't recieve a request in a set amount of time in order to save on resources, aswell as removing
+runtimes that are no longer needed.
+
+#### Flowchart Visualisation:
+
+![Flowchart Visualisation](flowchart.png)
+
```

Comment:
![image](https://user-images.githubusercontent.com/1297371/128515474-c804db15-11fa-4366-8193-92cb23f3ec47.png)
This part is also not correct. We don't necessarily need to start the runtime when a tag is activated. It's easier to delegate this responsibility to the first execution. We also don't need to stop the previous runtime. It can be stopped automatically after inactivity by the executor.